Add rusty wrapper of dtrace-sys

skywhale opened the pull request:
This adds a wrapper of `dtrace-sys` that provides more rusty / safe interface. There are a lot of limitations still, but I confirmed it is at least able to probe all the syscalls at the kernel and extract function names.

For https://github.com/tonarino/acoustic_profiler/issues/1

skywhale committed:
Add rusty wrapper of dtrace-sys.

skywhale committed:
Add missing files.

skywhale committed:
cargo fmt

goodhoko commented on crates/dtrace/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+    ///
+    /// Available DTrace options:
+    /// https://docs.oracle.com/en/operating-systems/solaris/oracle-solaris/11.4/dtrace-guide/consumer-options.html
+    pub fn execute_program(&self, program: &str, options: &[(&str, &str)]) -> Result<(), Error> {
```
It would be nice to encode the options within the type-system with eg. a `Vec<DtraceOption>` or `HashSet<DtraceOption>` where `DtraceOption` is an enum.

Though I don't think it's worth the effort now.

skywhale replied:
Yes, I thought about it, too. Let me leave a TODO as a something to come back later :)

goodhoko commented on crates/dtrace/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+    // informs about the state of the program execution.
+    //
+    // The duration it blocks depends on `switchrate`, `statusrate` and `aggrate` options used for
+    // the program execution.
```
```suggestion
    /// Blocks for some time and consumes the probe data collected by DTrace. The |ProgramStatus|
    /// informs about the state of the program execution.
    ///
    /// The duration it blocks depends on `switchrate`, `statusrate` and `aggrate` options used for
    /// the program execution.
```

skywhale replied:
Good catch :)

goodhoko commented on crates/dtrace/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+    pub fn wait_and_consume(
+        &mut self,
+        probes: &mut Vec<ProbeData>,
+    ) -> Result<ProgramStatus, Error> {
```
Returning the `Vec` instead of taking a mutable reference may be more ergonomic?

```suggestion
    pub fn wait_and_consume(&mut self) -> Result<(ProgramStatus, Vec<ProbeData>), Error> {
```

Instead of tuple a named struct could be used. Eg.

```suggestion
    struct ConsumedData {
        status: ProgramStatus,
        data: Vec<ProbeData>
    }

    pub fn wait_and_consume(&mut self) -> Result<ConsumedData, Error>
```

though there's definitely a better name for that struct. .)

goodhoko commented on crates/dtrace/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+            ProgramStatus::Ongoing,
+            dtrace.wait_and_consume(&mut probes)?
+        );
+        assert!(0 < probes.len());
```
super nit: I'd reverse the comparison to

```suggestion
        assert!(probes.len() > 0);
```

which IMO reads much closer to natural language "there's *more* than zero probes".

skywhale replied:
Addressed :)

goodhoko reviewed (approved):
Looks good to me! Just a few ideas mostly on the wrapper's ergonomics.

Generally, it'd be nice to pull functions' descriptions from Oracle docs to the doc-comments here. That would be especially nice if this wrapper is ever published. (I think it's a good candidate 😉)

skywhale committed:
Address Jen's review comments.

skywhale committed:
Cargo fmt

goodhoko reviewed (approved):
LGTM! (There's some lints failing)

skywhale commented:
Thank you @goodhoko for your review. I'm sorry for pickling this PR for so long. Now it tastes like kimchi 😛. I've addressed all of your comments and merging.
